enclave/core/sgx/calls.c

```diff
@@ -460,14 +461,23 @@ static void _handle_ecall(
             arg_out = oe_handle_call_enclave_function(arg_in);
             break;
         }
-        case OE_ECALL_DESTRUCTOR:
+        case OE_ECALL_CALL_EXIT_FUNCTIONS:
         {
-            /* Call functions installed by oe_cxa_atexit() and oe_atexit() */
-            oe_call_atexit_functions();
+            if (!__oe_exit_functions_done)
+            {
+                /* Call functions installed by oe_cxa_atexit() and oe_atexit()
+                 */
+                oe_call_atexit_functions();
 
-            /* Call all finalization functions */
-            oe_call_fini_functions();
+                /* Call all finalization functions */
+                oe_call_fini_functions();
 
+                __oe_exit_functions_done = 1;
+            }
+            break;
+        }
+        case OE_ECALL_DESTRUCTOR:
+        {
```

Comment:
We can refactor the code to `_oe_call_at_exit_functions` and call it both in `OE_ECALL_CALL_EXIT_FUNCTIONS` as well as `OE_ECALL_DESTRUCTOR`. That way the current behavior will be retained if ever existing hosts load a newer enclave.